### ISSUE TITLE
Create Hello World Examples

### ISSUE THREAD

**Issue by ghost** [ewaltuch/hello-w-make-example#2]

**Comment by ewaltuch** [comment]
Setting up a "buildroot-external" repository is a great way to manage your custom packages, configurations, and root filesystem overlays independently of the main Buildroot source tree. This approach allows you to maintain and reuse your customizations across different versions of Buildroot. Below are the steps to create and set up a `buildroot-external` repository.

### Step 1: Create the `buildroot-external` Repository Structure

First, create a new directory for your `buildroot-external` repository:

```bash
mkdir buildroot-external
cd buildroot-external
```

Inside this directory, you'll need to create the following structure:

```bash
mkdir -p package rootfs-overlay configs
```

Your directory structure should look like this:

```
buildroot-external/
├── package/
├── rootfs-overlay/
└── configs/
```

### Step 2: Set Up `external.mk`

The `external.mk` file tells Buildroot where to find your custom packages, rootfs overlay, and configuration files. Create this file in the root of your `buildroot-external` repository:

```bash
touch external.mk
```

Add the following content to `external.mk`:

```makefile
# Add custom packages to Buildroot
BR2_EXTERNAL_BUILDROOT_EXTERNAL_PATH := $(call qstrip,$(BR2_EXTERNAL_BUILDROOT_EXTERNAL_PATH))
BR2_EXTERNAL_BUILDROOT_EXTERNAL_PATH += $(DIR)/package

# Define the rootfs overlay directory
BR2_ROOTFS_OVERLAY := $(DIR)/rootfs-overlay
```

This file specifies the paths to your custom packages and rootfs overlay.

### Step 3: Create `external.desc`

The `external.desc` file provides a description for your `buildroot-external` repository. Create this file in the root of your `buildroot-external` repository:

```bash
touch external.desc
```

Add a brief description:

```text
Custom external repository for Buildroot, containing custom packages, rootfs overlay, and configurations.
```

### Step 4: Add Custom Packages

Inside the `package/` directory, you can create subdirectories for each of your custom packages. For instance, if you have the `hello-w-make-example` package, your structure would look like this:

```bash
mkdir -p package/hello-w-make-example
```

Place the `hello-w-make-example.mk` and `Config.in` files you created earlier into this directory.

### Step 5: Set Up Root Filesystem Overlay

In the `rootfs-overlay/` directory, create the directory structure that should be overlaid onto the root filesystem. For example:

```bash
mkdir -p rootfs-overlay/etc/init.d
```

Place any files you want to overlay onto the target root filesystem in their respective locations within this directory.

### Step 6: Create Configurations

You can place any custom Buildroot configuration files in the `configs/` directory. For example, if you have a custom Buildroot defconfig file:

```bash
cp /path/to/your/custom_defconfig configs/my_custom_defconfig
```

### Step 7: Use the `buildroot-external` Repository with Buildroot

When you're ready to use this external repository with Buildroot, follow these steps:

1. **Clone or navigate to your Buildroot directory:**

   ```bash
   cd /path/to/buildroot
   ```

2. **Run Buildroot with your external repository:**

   ```bash
   make BR2_EXTERNAL=/path/to/buildroot-external menuconfig
   ```

   In the `menuconfig` interface, you should see your external repository's packages and configurations available.

3. **Select your custom configurations or packages:**

   Use the `menuconfig` to select your custom defconfig, packages, or enable the rootfs overlay.

4. **Build your system:**

   ```bash
   make
   ```

### Step 8: Version Control Your `buildroot-external` Repository

To ensure your external repository is portable and can be used with any version of Buildroot, consider placing it under version control:

```bash
git init
git add .
git commit -m "Initial commit of buildroot-external repository"
```

You can now share this repository with others or use it across different projects and versions of Buildroot.

### Summary

By setting up a `buildroot-external` repository, you can efficiently manage your custom Buildroot packages, root filesystem overlays, and configurations independently of the main Buildroot source tree. This setup promotes reusability and modularity, allowing you to easily adapt your work across different versions of Buildroot.

**Comment by ewaltuch** [comment]
Certainly! To create a custom layer and recipe for the Yocto project that builds the software from the given Git repository (`git@github.com:eliwaltuch/hello-w-make-example.git`), follow the detailed steps below.

### Step 1: Set Up Your Yocto Environment

1. **Prepare the Yocto environment**: Ensure that you have a Yocto build environment already set up. If you do not have one set up, you can start with one of the Yocto Project releases like `poky` (e.g., `dunfell`, `kirkstone`, `langdale`).

2. **Initialize the Yocto build environment**: Run the following commands to initialize the environment:
    ```bash
    source /path/to/poky/oe-init-build-env
    ```

### Step 2: Create a New Layer

1. **Create a new layer**: Use the `bitbake-layers` script to create a new layer, let's name it `meta-hello`.

    ```bash
    bitbake-layers create-layer ../meta-hello
    ```

2. **Add the new layer to your build configuration**:
   
    Edit the `conf/bblayers.conf` file located in your `build` directory to include the new layer:

    ```bash
    BBLAYERS += " /path/to/meta-hello "
    ```

### Step 3: Create a New Recipe

1. **Create directories for the new recipe**:
   
    Navigate to the `meta-hello` directory and create the necessary directory structure for your recipe. Let's call the recipe `hello-world`:

    ```bash
    cd ../meta-hello
    mkdir -p recipes-example/hello-world
    ```

2. **Create the recipe file**:

    Create a new file named `hello-world_1.0.bb` under `recipes-example/hello-world`:

    ```bash
    touch recipes-example/hello-world/hello-world_1.0.bb
    ```

3. **Edit the recipe file**:

    Open `hello-world_1.0.bb` with your favorite text editor and add the following content:

    ```bitbake
    SUMMARY = "Hello World application using Makefile"
    DESCRIPTION = "An example Hello World application that uses a Makefile for building"
    LICENSE = "MIT"
    LIC_FILES_CHKSUM = "file://${COMMON_LICENSE_DIR}/MIT;md5=0835ec6d0f0605da980b8b667f5f08a1"

    SRC_URI = "git://github.com/eliwaltuch/hello-w-make-example.git;protocol=ssh"
    SRCREV = "master"  # Adjust if you want to build a specific revision

    # Set the branch to fetch from the repository
    PV = "1.0+git${SRCPV}"
    S = "${WORKDIR}/git"

    # Specify the tasks for building
    inherit pkgconfig autotools

    do_compile() {
        oe_runmake
    }

    do_install() {
        install -d ${D}${bindir}
        install -m 0755 hello ${D}${bindir}/hello
    }

    # Define where the hello executable should be installed in the target rootfs
    FILES_${PN} += "${bindir}/hello"
    ```

### Step 4: Build the Recipe

1. **Run `bitbake` to build the recipe**:

    Go back to your build directory and run:

    ```bash
    bitbake hello-world
    ```

2. **Check the output**:

    If everything is set up correctly, BitBake should fetch the source code from the specified Git repository, build the `hello` executable, and create the package.

### Step 5: Test the Built Package

1. **Add the package to your image**:

    To include the `hello-world` application in your final image, add the package to your image by editing `conf/local.conf`:

    ```bash
    IMAGE_INSTALL:append = " hello-world"
    ```

2. **Rebuild the image**:

    Rebuild your image to include the new package:

    ```bash
    bitbake core-image-minimal
    ```

3. **Deploy and Test**:

    Deploy the image to your target device and test to ensure that the `hello` application is correctly installed and working.

### Conclusion

By following these steps, you've successfully created a custom layer and recipe for the Yocto Project that builds the Hello World application from the given Git repository. You can now customize the recipe further or extend the layer with additional recipes as needed.
